reinforcement_learning/rlclientlib/ranking_event.cc

```diff
@@ -56,79 +55,133 @@ namespace reinforcement_learning {
+    : event(event_id, pass_prob)
+  {
+    _context = std::string(context);
+    for (auto const &r : response) {
+      _a_vector.push_back(r.action_id);
+      _p_vector.push_back(r.probability);
+    }
+    _model_id = std::string(response.get_model_id());
+  }
+
   ranking_event::ranking_event(ranking_event&& other)
     : event(std::move(other))
     , _body(std::move(other._body))
-  {}
+    , _context(std::move(other._context))
+    , _a_vector(std::move(other._a_vector))
+    , _p_vector(std::move(other._p_vector))
+    , _model_id(std::move(other._model_id))
+  { }
 
   ranking_event& ranking_event::operator=(ranking_event&& other) {
     if (&other != this) {
       event::operator=(std::move(other));
       _body = std::move(other._body);
+      _context = std::move(other._context);
+      _a_vector = std::move(other._a_vector);
+      _p_vector = std::move(other._p_vector);
+      _model_id = std::move(other._model_id);
     }
     return *this;
   }
 
-  void ranking_event::serialize(u::data_buffer& oss) {
-    oss << _body;
-    if (_pass_prob < 1) {
-      oss << R"(,"pdrop":)" << (1 - _pass_prob);
-    }
-    oss << R"(})";
+  flatbuffers::Offset<RankingEvent> ranking_event::serialize_eventhub_message(flatbuffers::FlatBufferBuilder& builder) {
+    short version = 1;
+    auto event_id_offset = builder.CreateString(_event_id);
+
+    auto a_vector_offset = builder.CreateVector(_a_vector);
+    auto p_vector_offset = builder.CreateVector(_p_vector);
+
+    auto context_offset = builder.CreateString(_context);
+
+    auto vw_state_offset = VW::Events::CreateVWStateType(builder, builder.CreateString(_model_id));
+
+    return VW::Events::CreateRankingEvent(builder, version, event_id_offset, a_vector_offset, context_offset, p_vector_offset, vw_state_offset);
   }
 
   ranking_event ranking_event::choose_rank(u::data_buffer& oss, const char* event_id, const char* context,
     unsigned int flags, const ranking_response& resp, float pass_prob) {
 
     //add version and eventId
     oss << R"({"Version":"1","EventId":")" << event_id << R"(")";
+
     if (flags & action_flags::DEFERRED) {
       oss << R"(,"DeferredAction":true)";
     }
+
     //add action ids
     oss << R"(,"a":[)";
-    if ( resp.size() > 0 ) {
-      for ( auto const &r : resp )
+    if (resp.size() > 0) {
+      for (auto const &r : resp)
         oss << r.action_id + 1 << ",";
       oss.remove_last();//remove trailing ,
     }
 
     //add probabilities
     oss << R"(],"c":)" << context << R"(,"p":[)";
-    if ( resp.size() > 0 ) {
-      for ( auto const &r : resp )
+    if (resp.size() > 0) {
+      for (auto const &r : resp)
         oss << r.probability << ",";
       oss.remove_last();//remove trailing ,
     }
 
     //add model id
     oss << R"(],"VWState":{"m":")" << resp.get_model_id() << R"("})";
-
     return ranking_event(event_id, pass_prob, oss.str());
-	}
+  }
+
+  void ranking_event::serialize(u::data_buffer& oss) {
```

Comment:
Merge conflict: This method is not used anymore. On the other side there is no pdrop support in the new serialization.

Reply:
I will double check that. I ported the event schema from previous event_test.cc. If it was not tested there, it will likely be missed since we didn't have a schema before. Is it float?

---
In reply to: [228608998](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1653#discussion_r228608998) [](ancestors = 228608998)